Paper: Pandata -- The Scalable Open-Source Analysis Stack

jbednar opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@deniederhut<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@LMedinaU, @tkoyama010<!--end-reviewers-list-->

jbednar committed:
Initial draft

martindurant committed:
comments

jbednar committed:
Added edits by droumis

jbednar committed:
Updated based on co-author comments

jbednar committed:
Trigger CI

deniederhut commented:
@scoobies assign @deniederhut as editor

scoobies commented:
Assigned! @deniederhut is now the editor

deniederhut commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1038/s41586-020-2649-2 is OK
- 10.1109/MCSE.2007.55 is OK

MISSING DOIs

- 10.25080/majora-92bf1922-00a may be a valid DOI for title: Data Structures for Statistical Computing in Python

INVALID DOIs

- None
```

jbednar committed:
Added missing DOI

deniederhut commented:
@scoobies add @LMedinaU to reviewers

scoobies commented:
@LMedinaU added to the reviewers list!

deniederhut commented:
@scoobies code of conduct

scoobies commented:
Our CoC: https://www.scipy2023.scipy.org/code-of-conduct

scoobies commented:
Reports of abusive or harassing behavior may be reported to scipy@enthought.com

deniederhut commented:
@scoobies add @tkoyama010 to reviewers

scoobies commented:
@tkoyama010 added to the reviewers list!

LMedinaU commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,166 @@
+Other relevant areas that need work include the development of domain-general tools for capturing metadata, conventions, consistent units, etc. per domain (like the CF conventions for climate science). Furthermore, improved tools for collaboration are needed, such as implementing parts of the JupyterHub roadmap (jupyterhub.readthedocs.io/en/stable/contributing/roadmap.html). Lastly, better docume ...
+
+
+\subsubsection{Conclusion}
```
Please use the same subtitle format as the other subtitle (e.g. Introduction). This suggestion is based on the pdf version.

jbednar replied:
```suggestion
\subsection{Conclusion}
```

LMedinaU commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,166 @@
+
+There are also full alternative stacks to consider, such as tools like Hadoop and Spark from the Apache Foundation. Most Apache tools rely on the Java Virtual Machine (JVM) that provides OS-independent computation but requires a heavyweight runtime compared to the SOSA tools, making it awkward to combine most Apache tools with SOSA tools. SOSA already offers flexible support for distributed compu ...
+
+\subsubsection{Future work}
```
Please use the same subtitle format as the other subtitle (e.g. Introduction)

jbednar replied:
```suggestion
\subsection{Future work}
```

LMedinaU commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,166 @@
+
+Whether or not any domain-specific processing has been done, the raw and processed data then need to be visualized so that humans can understand and validate them. Most Python visualization tools (see pyviz.org for a complete list) have limitations that prevent them from being suitable for the SOSA stack. Many of them are limited to relatively small amounts of data, lack support for the various d ...
+
+\subsubsection{UI}
```
Suggestion: Title as User Interface, instead of UI, to be consistent with the previous subtitle’s format you are using.

jbednar replied:
```suggestion
\subsubsection{User Interface (UI)}
```

LMedinaU commented on papers/james_bednar/metadata.yaml:
```diff
@@ -0,0 +1,19 @@
+---
+abstract:
+   - As the scale of scientific data analysis continues to grow, traditional domain-specific tools often struggle to manage the increasing size and complexity. These tools also face sustainability challenges due to a relatively narrow user base, a limited pool of contributors, and constrained funding sources. We introduce the Scalable Open-Source Analysis (SOSA) stack as a solution, emphasizing t ...
```
Overall, the abstract provides a clear overview of the paper's focus and contributions. The SOSA stack seems to address important issues in scientific data analysis, including scalability, sustainability, and interoperability. The integration of domain-independent tools while preserving the depth of domain-specific analyses is a promising approach.

tkoyama010 commented on papers/james_bednar/sosa-finance.png:
~I am a little uncomfortable with the layout of "Domain-specific visualization" character.~
Sorry, I got that Domain-specific and Visualizaition are separated and Domain-specific is empty.

tkoyama010 commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,166 @@
+
+\begin{figure*}[t]
+    \noindent\makebox[\textwidth][c]{\includegraphics[width=\textwidth]{sosa.png}}
+    \caption{SOSA: The Scalable Open-Source Analysis Stack\label{sosa}. Many data-processing problems can be solved by choosing one or more items from each of these categories (described later in the text). The tools shaded in gray are not part of SOSA itself but are used together with SOSA to make a complete stack suitable for a particular research domain.}
```
I felt the title was a little too long. How about the move "Many data-processing problems..." to the footnotes?

tkoyama010 commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,166 @@
+
+Library authors have to choose between different designs and alternative technologies when deciding which features are implemented and how they are built.  Here we will unpack each of the above considerations driving the SOSA collection of libraries and examine how these libraries implement those principles.
+
+\subsubsection{Domain independent} SOSA tools are all designed to be applied across scientific domains, without making design choices that constrain them to a narrow subfield or topic. At first it may seem that having all your tools be written specifically for your particular research area would be ideal, but we argue that it is neither desirable nor feasible to draw strict boundaries between act ...
```
We suggest aligning the layout :)
```suggestion
\subsubsection{Domain independent}
SOSA tools are all designed to be applied across scientific domains, without making design choices that constrain them to a narrow subfield or topic. At first it may seem that having all your tools be written specifically for your particular research area would be ideal, but we argue that it is neither desirable nor feasible to draw strict boundaries between active scientific research areas. Should you have to switch to an entirely different stack whenever you do collaborative work? Or if you want to do something novel, differing slightly from assumptions previously made in your research area? A deep, difficult-to-change stack that encodes a fixed and unnecessarily brittle model of your domain will make scientific progress difficult to achieve. Of course, not every aspect of research crosses multiple domains or challenges core assumptions, but using SOSA tools ensures that doing so will come at relatively little cost, because the same tools are valid across a very wide range of fields. SOSA tools can also draw from expertise and funding across all of science and analysis, rather than from a single narrow area.
```

tkoyama010 commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,166 @@
+
+\subsubsection{Domain independent} SOSA tools are all designed to be applied across scientific domains, without making design choices that constrain them to a narrow subfield or topic. At first it may seem that having all your tools be written specifically for your particular research area would be ideal, but we argue that it is neither desirable nor feasible to draw strict boundaries between act ...
+
+\subsubsection{Efficient} Because Python itself is a relatively slow, interpreted language, native Python code is often suitable only for relatively small problems, limiting the audience for any tool built purely in Python with Python dependencies. To make efficient use of the computational resources available, SOSA libraries either wrap fast code written in compiled languages like FORTRAN and C, ...
```
```suggestion
\subsubsection{Efficient}
Because Python itself is a relatively slow, interpreted language, native Python code is often suitable only for relatively small problems, limiting the audience for any tool built purely in Python with Python dependencies. To make efficient use of the computational resources available, SOSA libraries either wrap fast code written in compiled languages like FORTRAN and C, or they use the Numba Python JIT compiler to compile Python syntax into machine code that runs as fast as compiled C or Rust or FORTRAN. Custom domain-specific code for SOSA can easily use Numba as well, so that the overall workflow and pipeline is not limited by the speed of Python.
```

tkoyama010 replied:
```suggestion
\subsubsection{Efficient} Because Python itself is a relatively slow, interpreted language, native Python code is often suitable only for relatively small problems, limiting the audience for any tool built purely in Python with Python dependencies. To make efficient use of the computational resources available, SOSA libraries either wrap fast code written in compiled languages like FORTRAN and C, or they use the Numba Python JIT compiler to compile Python syntax into machine code that runs as fast as compiled C or Rust or FORTRAN. Custom domain-specific code for SOSA can easily use Numba as well, so that the overall workflow and pipeline are not limited by the speed of Python.
```

tkoyama010 commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,166 @@
+
+\subsubsection{Efficient} Because Python itself is a relatively slow, interpreted language, native Python code is often suitable only for relatively small problems, limiting the audience for any tool built purely in Python with Python dependencies. To make efficient use of the computational resources available, SOSA libraries either wrap fast code written in compiled languages like FORTRAN and C, ...
+
+\subsubsection{Scalable} Even compiled code is not sufficient to address the largest problems, which require more memory than is available on a single machine, or require long computation that is feasible only when split across many processors working simultaneously. To cover all these cases, SOSA tools support distributed computation using Dask on hundreds or thousands of processors, while also  ...
```
```suggestion
\subsubsection{Scalable}
Even compiled code is not sufficient to address the largest problems, which require more memory than is available on a single machine, or require long computation that is feasible only when split across many processors working simultaneously. To cover all these cases, SOSA tools support distributed computation using Dask on hundreds or thousands of processors, while also fully supporting a single laptop or workstation. That way there is no need to switch to a different stack when you hit a problem larger than your machine, or conversely when you want to work on small problems that do not need extensive infrastructure. Supporting distributed computation efficiently requires tooling support at every level, starting with chunked binary file formats like Parquet and Zarr that let each processor access data independently, and culminating in visualization tools like Datashader that can render plots in separate chunks on each processor and combine them for the final display.
```

tkoyama010 commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,166 @@
+\subsubsection{Scalable} Even compiled code is not sufficient to address the largest problems, which require more memory than is available on a single machine, or require long computation that is feasible only when split across many processors working simultaneously. To cover all these cases, SOSA tools support distributed computation using Dask on hundreds or thousands of processors, while also  ...
+
+\subsubsection{Cloud friendly}
+The enormous datasets now available in many research areas that are hosted in a remote data center can be prohibitively expensive (in time and money) to download locally. If you only need a small part of the data, SOSA tools support efficient access to the relevant chunks of each remote file without having to download it all first, using fsspec to provide uniform filesystem-like access for data o ...
```
initating -> initiating
```suggestion
The enormous datasets now available in many research areas that are hosted in a remote data center can be prohibitively expensive (in time and money) to download locally. If you only need a small part of the data, SOSA tools support efficient access to the relevant chunks of each remote file without having to download it all first, using fsspec to provide uniform filesystem-like access for data on local disk, cloud storage, web servers, and many other locations. If the total number of chunks that you need to access is still too high to download locally, SOSA tools also support remote computation, with the processing done on a remote cloud-computing server with a high-bandwidth connection to the data server, while providing a user interface on your local machine. A typical cloud-computing configuration for SOSA would be to have a remote JupyterHub system with Dask installed running on the cloud server close to the data. Users then contact the remote system from their local browser, initiating a remote session for computation and data access but with interactivity on the local machine using web-based controls and visuals. Using Datashader on the remote system provides server-side rendering so that only a rendered image of the data ever need be transferred across low-bandwidth connections, providing visualization without having to transfer large datasets.
```

tkoyama010 commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,166 @@
+When approaching a particular task, a researcher can either choose a monolithic tool that addresses all aspects of the task, or they can combine lower-level tools that together accomplish the task. Monolithic tools are attractive when they fully cover a specific use case, but given the dynamic nature of research, it is unreasonable to expect a monolithic tool to cover all the requirements of a pa ...
+
+\subsubsection{Visualizable}
+For computing tasks that complete in the background without any visual output, it is easy to ensure that they are scriptable and cloud friendly. However, doing good science requires understanding each of the processing steps in complete detail, and if there are any unobservable black boxes in your workflows, that is where bugs will hide. To make sure that the work is being done correctly, it is c ...
```
complete -> are complete
```suggestion
For computing tasks that are complete in the background without any visual output, it is easy to ensure that they are scriptable and cloud friendly. However, doing good science requires understanding each of the processing steps in complete detail, and if there are any unobservable black boxes in your workflows, that is where bugs will hide. To make sure that the work is being done correctly, it is crucial to be able to represent each of the computing steps involved in a way that a human can easily grasp what is happening, with a minimum of extra effort that discourages exploration. Often, a remote computing job will export data that is then subsampled and downloaded locally for analysis, but any step that adds friction and covers up the raw data makes it more likely that important issues and insights will be missed. Accordingly, the HoloViz tools included in the SOSA stack are designed to make the full set of data available at any point in the computation, by supporting efficient interactive visualization of even distributed or GPU-based data of any size using Datashader and hvPlot, assembled from separately computed chunks for display on any device.
```

jbednar replied:
```suggestion
For computing tasks that operate in the background without any visual output, it is easy to ensure that they are scriptable and cloud friendly. However, doing good science requires understanding each of the processing steps in complete detail, and if there are any unobservable black boxes in your workflows, that is where bugs will hide. To make sure that the work is being done correctly, it is crucial to be able to represent each of the computing steps involved in a way that a human can easily grasp what is happening, with a minimum of extra effort that discourages exploration. Often, a remote computing job will export data that is then subsampled and downloaded locally for analysis, but any step that adds friction and covers up the raw data makes it more likely that important issues and insights will be missed. Accordingly, the HoloViz tools included in the SOSA stack are designed to make the full set of data available at any point in the computation, by supporting efficient interactive visualization of even distributed or GPU-based data of any size using Datashader and hvPlot, assembled from separately computed chunks for display on any device.
```

LMedinaU commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,166 @@
+\subsubsection{Environments and reproducibility}
+
+A final category of tools underlying all the rest shown in figure \ref{sosa} centers on how these libraries are packaged and put together into Python environments for solving any particular problem. Many different options are available for Python package and environment management, but SOSA tools are typically used with the conda package and environment manager, because it tracks binary dependenc ...
+https://www.anaconda.com/blog/8-levels-of-reproducibility for details).
```
Please, check this website (page not found).

jbednar replied:
Maybe it was briefly offline? Testing https://www.anaconda.com/blog/8-levels-of-reproducibility works fine for me at the moment on different computers and with different people logged in.

tkoyama010 replied:
It works fine for me too.

LMedinaU reviewed (commented):
In general, this paper is clearly written and well-structured, successfully achieving its overall objective of introducing an open-source alternative aimed at providing a framework that integrates a diverse set of Python libraries for general data processing. The attributes of the framework are presented coherently and supported by technical arguments. To further enhance the quality of the final version of the paper, I would like to offer the following suggestions:

•	The introduction could benefit from citing additional references to support the context developed in this section. Here are a few suggestions that may be worth exploring:

Sagiroglu, S., & Sinanc, D. (2013, May). Big data: A review. In 2013 international conference on collaboration technologies and systems (CTS) (pp. 42-47). IEEE.

Verbraeken, J., Wolting, M., Katzy, J., Kloppenburg, J., Verbelen, T., & Rellermeyer, J. S. (2020). A survey on distributed machine learning. Acm computing surveys (csur), 53(2), 1-33.

Sudha, P., & Gunavathi, R. (2016). A survey paper on map reduce in big data. International Journal of Science and Research, 5(9).

Stančin, I., & Jović, A. (2019, May). An overview and comparison of free Python libraries for data mining and big data analysis. In 2019 42nd International convention on information and communication technology, electronics and microelectronics (MIPRO) (pp. 977-982). IEEE.

Zaharia, Matei, et al. "Apache spark: a unified engine for big data processing." Communications of the ACM 59.11 (2016): 56-65.

Please note that these suggestions are not exhaustive, and it is recommended to conduct a thorough literature review to identify relevant references that best align with the context and objectives of your paper.

•	Please also ensure that you maintain consistency in the reference format throughout the text (e.g., using the IEEE format) when citing websites as examples.holoviz.org. Similarly, when mentioning alternative tools like Ray, Vaex, Polars, VegaFusion, Hadoop, and Spark, please follow the same reference format.

deniederhut commented:
@scoobies mark pending comment

tkoyama010 commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,166 @@
+The enormous datasets now available in many research areas that are hosted in a remote data center can be prohibitively expensive (in time and money) to download locally. If you only need a small part of the data, SOSA tools support efficient access to the relevant chunks of each remote file without having to download it all first, using fsspec to provide uniform filesystem-like access for data o ...
+
+\subsubsection{Multi-architecture}
+Because researchers typically use Windows or Mac systems locally, while cloud servers typically run Linux, research code needs to be independent of the operating system for it to be equally well supported on local and remote systems. A software stack tied to a particular OS not only shuts out users who are not on that OS, it often implicitly favors either cloud or local usage for that stack, furt ...
```
We can remove "they"
```suggestion
Because researchers typically use Windows or Mac systems locally, while cloud servers typically run Linux, research code needs to be independent of the operating system for it to be equally well supported on local and remote systems. A software stack tied to a particular OS not only shuts out users who are not on that OS, it often implicitly favors either cloud or local usage for that stack, further reducing the community size and range of problems that can be addressed by a particular stack. SOSA tools are all either fully OS independent or support Linux (Intel or ARM), Mac (Intel or M1/M2), and Windows. Similarly, some problems can be addressed orders of magnitude faster on a general-purpose graphics processing unit (GPU) than on a conventional CPU, yet many researchers do not have access to GPUs or are working on problems unsuited for them, making it essential that a general-purpose software stack support both GPU and CPU usage as appropriate. Many SOSA tools provide support for GPUs where appropriate.
```

jbednar committed:
Update papers/james_bednar/james_bednar.tex

Thanks for all the updates! Applying them in batch isn't working, so get ready for a bunch of individual commits!

Co-authored-by: Tetsuo Koyama <tkoyama010@gmail.com>

jbednar committed:
Update papers/james_bednar/james_bednar.tex

Co-authored-by: Tetsuo Koyama <tkoyama010@gmail.com>

jbednar committed:
Update papers/james_bednar/james_bednar.tex

Co-authored-by: Tetsuo Koyama <tkoyama010@gmail.com>

jbednar committed:
Update papers/james_bednar/james_bednar.tex

Co-authored-by: Tetsuo Koyama <tkoyama010@gmail.com>

jbednar committed:
Update papers/james_bednar/james_bednar.tex

Co-authored-by: Tetsuo Koyama <tkoyama010@gmail.com>

jbednar committed:
Update papers/james_bednar/james_bednar.tex

jbednar committed:
Update papers/james_bednar/james_bednar.tex

jbednar committed:
Update papers/james_bednar/james_bednar.tex

jbednar committed:
Update papers/james_bednar/james_bednar.tex

jbednar commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,169 @@
+SOSA tools are all designed to be applied across scientific domains, without making design choices that constrain them to a narrow subfield or topic. At first it may seem that having all your tools be written specifically for your particular research area would be ideal, but we argue that it is neither desirable nor feasible to draw strict boundaries between active scientific research areas. Shou ...
+
+\subsubsection{Efficient}
+Because Python itself is a relatively slow, interpreted language, native Python code is often suitable only for relatively small problems, limiting the audience for any tool built purely in Python with Python dependencies. To make efficient use of the computational resources available, SOSA libraries either wrap fast code written in compiled languages like FORTRAN and C, or they use the Numba Pyt ...
```
```suggestion
Because Python itself is a relatively slow, interpreted language, native Python code is often suitable only for relatively small problems, limiting the audience for any tool built purely in Python with Python dependencies. To make efficient use of the computational resources available, SOSA libraries either wrap fast code written in compiled languages like FORTRAN and C, or they use the Numba Python JIT compiler to compile Python syntax into machine code that runs as fast as compiled C or Rust or FORTRAN. Custom domain-specific code for SOSA can easily use Numba as well, so that the overall workflows and pipelines are not limited by the speed of Python.
```

jbednar committed:
Update papers/james_bednar/james_bednar.tex

tkoyama010 commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,169 @@
+
+\subsection{Examples}
+
+To get an understanding of how these tools fit together, there is a variety of example workflows based on SOSA tools available online. The website examples.holoviz.org includes a variety of research topics across domains, with SOSA tools featured prominently in most of them and particularly the Attractors, Census, Ship traffic, and Landsat examples. The holoviz.org site has a tutorial that brings ...
```
```suggestion
To understand how these tools fit together, there are a variety of example workflows based on SOSA tools available online. The website examples.holoviz.org includes a variety of research topics across domains, with SOSA tools featured prominently in most of them and particularly the Attractors, Census, Ship traffic, and Landsat examples. The holoviz.org site has a tutorial that brings in many of the SOSA packages to solve an example research problem. The Pangeo Project Gallery at pangeo.io and Project Pythia at projectpythia.org both include detailed examples in the specific area of climate science. Each SOSA package also has its own documentation and website illustrating what it can do, often in combination with other SOSA tools, such as the Dask examples at stories.dask.org.
```

jbednar replied:
```suggestion
To understand how these tools fit together, a variety of example workflows based on SOSA tools can be found online. The website examples.holoviz.org includes research topics in different domains, with SOSA tools featured prominently in most of them and particularly the Attractors, Census, Ship Traffic, and Landsat examples. The holoviz.org site has a tutorial that brings in many of the SOSA packages to solve an example research problem. The Pangeo Project Gallery at pangeo.io and Project Pythia at projectpythia.org both include detailed examples in the specific area of climate science. Each SOSA package also has its own documentation and website illustrating what it can do, often in combination with other SOSA tools, such as the Dask examples at stories.dask.org.
```

tkoyama010 commented:
The review is over. Thank you for a great paper!

jbednar committed:
Update papers/james_bednar/james_bednar.tex

deniederhut commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1038/s41586-020-2649-2 is OK
- 10.25080/majora-92bf1922-00a is OK
- 10.1109/MCSE.2007.55 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

deniederhut commented:
> 10.25080/majora-92bf1922-00a is OK

Always love to see SciPy papers in the reference section ♥️

deniederhut commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,169 @@
+% Up to 8 pages formatted
+\subsection{Introduction \label{introduction}}
+
+Modern science, engineering, and analysis rely on computational tools for working with data, such as the foundational NumPy \cite{harris:nature20}, Pandas \cite{mckinney:scipy10}, and Matplotlib \cite{hunter:cse07} libraries for Python. Over the past few decades, many different research areas and communities have built their own ``stacks'', i.e. layered sets of software tools that are combined to ...
```
I would expect to see citations for numpy and jupyter in this paper.

jbednar replied:
There was already a numpy citation, but I added one for jupyter (and many others) now.

deniederhut commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,169 @@
+Library authors have to choose between different designs and alternative technologies when deciding which features are implemented and how they are built.  Here we will unpack each of the above considerations driving the SOSA collection of libraries and examine how these libraries implement those principles.
+
+\subsubsection{Domain independent}
+SOSA tools are all designed to be applied across scientific domains, without making design choices that constrain them to a narrow subfield or topic. At first it may seem that having all your tools be written specifically for your particular research area would be ideal, but we argue that it is neither desirable nor feasible to draw strict boundaries between active scientific research areas. Shou ...
```
I am 100% sympathetic to this point of view, but not all of these tools scale well. Pandas famously does not! This is the motivation for dask dataframes, no?

jbednar replied:
Yes, that's tricky to be concrete about. A Dask dataframe is just a bag of Pandas dataframes, and so Pandas does scale well if used via Dask. It's not as if Dask has done away with the need for Pandas; it just makes it scalable. Similarly for Matplotlib -- Matplotlib itself isn't scalable, but when used with Datashader and a Dask dataframe, it can display server-side-rendered plots where the data never needs to leave the distributed set of servers.  I'll try to clarify that, but it's a pretty subtle distinction to convey.

deniederhut commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,169 @@
+SOSA tools are all designed to be applied across scientific domains, without making design choices that constrain them to a narrow subfield or topic. At first it may seem that having all your tools be written specifically for your particular research area would be ideal, but we argue that it is neither desirable nor feasible to draw strict boundaries between active scientific research areas. Shou ...
+
+\subsubsection{Efficient}
+Because Python itself is a relatively slow, interpreted language, native Python code is often suitable only for relatively small problems, limiting the audience for any tool built purely in Python with Python dependencies. To make efficient use of the computational resources available, SOSA libraries either wrap fast code written in compiled languages like FORTRAN and C, or they use the Numba Pyt ...
```
Cython?

jbednar replied:
People can use whatever libraries they like with SOSA, and Pandas itself uses cython, so I suppose I ought to mention it for completeness. Still, I wouldn't steer people writing new code to fit with these frameworks to cython; it doesn't offer any immediate additional benefit compared to what's in SOSA already and each new library mentioned just makes it more overwhelming for users to sort out...

jbednar replied:
Mentions Cython now.

deniederhut commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,169 @@
+Because Python itself is a relatively slow, interpreted language, native Python code is often suitable only for relatively small problems, limiting the audience for any tool built purely in Python with Python dependencies. To make efficient use of the computational resources available, SOSA libraries either wrap fast code written in compiled languages like FORTRAN and C, or they use the Numba Pyt ...
+
+\subsubsection{Scalable}
+Even compiled code is not sufficient to address the largest problems, which require more memory than is available on a single machine, or require long computation that is feasible only when split across many processors working simultaneously. To cover all these cases, SOSA tools support distributed computation using Dask on hundreds or thousands of processors, while also fully supporting a single ...
```
> To cover all these cases, SOSA tools support distributed computation using Dask on hundreds or thousands of processors, while also fully supporting a single laptop or workstation

SOSA stack? Or DASK stack?

jbednar replied:
We did discuss naming the stack after Dask, but in the end decided that because Dask is always optional, it would be confusing to do so. The strength of SOSA is if you use these particular tools that support Dask but don't require it, it is easy to support fully scalable operation without your users having to switch anything besides the Data API library. So if someone is developing for small problems only, we don't want to force them to use Dask; they can happily use Pandas natively and everything just works; they'll just need to pass in a Dask dataframe later if they encounter a larger problem. So in the end we decided it wasn't the Dask stack, it was the Dask- _supporting_ stack, and the reason to support Dask is to allow scalability, hence the Scalable Open-source Analysis stack.  Lots of tradeoffs in any naming exercise, of course!

deniederhut commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,169 @@
+Even compiled code is not sufficient to address the largest problems, which require more memory than is available on a single machine, or require long computation that is feasible only when split across many processors working simultaneously. To cover all these cases, SOSA tools support distributed computation using Dask on hundreds or thousands of processors, while also fully supporting a single ...
+
+\subsubsection{Cloud friendly}
+The enormous datasets now available in many research areas that are hosted in a remote data center can be prohibitively expensive (in time and money) to download locally. If you only need a small part of the data, SOSA tools support efficient access to the relevant chunks of each remote file without having to download it all first, using fsspec to provide uniform filesystem-like access for data o ...
```
To be clear -- NumPy and Pandas alone don't do this? Is this a dask-specific argument?

jbednar replied:
I'll try to clarify; for a single-CPU problem Pandas+Datashader is perfectly cloud friendly in that data never needs to be transmitted over the wire, it's just not scalable.  Matplotlib is cloud friendly if rendering to a PNG or SVG on the remote server. Bokeh alone is not cloud friendly for large data running remotely because it requires all the data to be sent to the local client for display. HoloViews+Datashader+Bokeh is cloud friendly since the data never needs to be sent all the way to the local client.  So many combinations!  I've tried to focus on the principles here, of leaving data where it lies and processing it there; detailing all the combinations that do or do not work well would be infeasible here. I'll see what I can do!

jbednar replied:
Clarified.

deniederhut commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -0,0 +1,169 @@
+The enormous datasets now available in many research areas that are hosted in a remote data center can be prohibitively expensive (in time and money) to download locally. If you only need a small part of the data, SOSA tools support efficient access to the relevant chunks of each remote file without having to download it all first, using fsspec to provide uniform filesystem-like access for data o ...
+
+\subsubsection{Multi-architecture}
+Because researchers typically use Windows or Mac systems locally, while cloud servers typically run Linux, research code needs to be independent of the operating system for it to be equally well supported on local and remote systems. A software stack tied to a particular OS not only shuts out users who are not on that OS, it often implicitly favors either cloud or local usage for that stack, furt ...
```
One day we will have the year of linux on the desktop, I swear

jbednar replied:
This is the year.

deniederhut reviewed (commented):
This feels like a position paper, but after looking through it, I'm not really clear what the position is. It's okay to have strong opinions, and scipy is a great place to say something daring among friends! But like... what is your stake here? 

I see articles in my news feed every day about polars and duckdb. I would love to see a pythonland response to these. Is this what you want this paper to be? Or am I misunderstanding your intention here?

tkoyama010 commented:
@deniederhut 
Good point.
Personally, I was reviewing this paper thinking that its main purpose was to classify and analyze packages; it would be valuable to provide the reader with a map of OSS packages.
In any case, let's hear @jbednar's thoughts.

jbednar commented:
I'm not sure what you mean by a response to duckdb and polars. The status of duckdb wrt this particular set of packages is that it is becoming well supported for use in combination with the others, because it meets a specific use case not already accounted for by the other packages in SOSA. There is not yet significant support for polars by these other packages, so for now polars is not part of SOSA, but there is definitely interest and I can imagine adding polars at some point. The point of this paper is to list the packages that currently support being used together to achieve the properties listed in the paper (scalability, interactivity, etc.), to advertise that such a collection of tools is available and adds up to form a coherent basis on which to build a complete scientific workflow. It is not meant to advertise for (or against) any one package.

jbednar committed:
Revised as suggested by reviewers

jbednar commented:
Ok, I think I've addressed the various reviewer comments; thanks, all! I added lots of references, but there could be dozens more to add since so many libraries are mentioned. I focused on adding references for each library that was extensively discussed. I also left various URLs in the main text rather than in the reference list if they were explicitly being used as URLs for someone to read and notice (e.g. `pangeo.io`) rather than simply as hyperlinks. They could all be moved to the bibliography, but I don't think that will do the readers a service in most cases.

deniederhut commented:
Thanks for the explanation! Okay so it sounds like I misunderstood. On second reading, this line right here jumped out at me:

> Each project is designed to address pain points being experienced by those collaborators, and the staff involved help the projects work together to add up to complete solutions.

There is a sort of implicit statement here that consulting is hard work; the specific problem domain varies a lot; and, after doing this for a while Anaconda realized that many of the problems each individual consulting project had could be abstracted away into more domain-general tools that could be reused (and maybe also that this gave the rest of the team more time/space/energy to focus on the domain-specific parts?). And not just reused by Anaconda! The person reading this paper can also build a successful company / research lab / government program using these tools.

Perhaps you disagree, but this seems to me like a valuable thing to make explicit. It's not just that these tools are open-source or scalable (although that is important) -- it's that they are battle tested.

jbednar committed:
Addressed additional reviewer comments

jbednar commented:
Thanks for that thoughtful commentary. I'm hesitant to promote Anaconda too much in a scientific forum, but yes, the business processes involved are crucial for this outcome, since any one researcher is normally only in their own domain, with groups like ours that cross all domains being rare. We do explicitly make a point that in our consulting projects we are able to spend more time focusing on the domain-specific work because we can build on all the previous general work. Plus it keeps everyone on my team happier -- we all love working on new and interesting problems, rather than serially re-solving the same problem with different names and different customers. :-) I've expanded that background in the text, now including the word "battle". :-)

martindurant commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -1,17 +1,17 @@
 Many of the components of these stacks date back decades before Python became popular, capturing important functionality but inheriting technical complexity and limitations that may no longer apply. For instance, domain-specific visualization and user interface (UI) tools are often tied to a local desktop operating system and graphical user interface (GUI), limiting the stack to working with data ...
 
-Could there be a better way? Yes! Today's Python ecosystem includes general data-processing tools that address tasks that cross research areas, domains, and industries, with flexible tools for storing, reading, processing, plotting, analyzing, modeling, and exploring data of any kind. This paper introduces a collection of such tools called SOSA: the Scalable Open-Source Analysis Stack.
+Could there be a better way? Yes! Today's Python ecosystem includes general data-processing tools that address tasks that cross research areas, domains, and industries, with flexible tools for storing, reading, processing, plotting, analyzing, modeling, and exploring data of any kind. This paper introduces a collection of such tools called SOSA: the Scalable Open-Source Analysis Stack, with ``sca ...
```
Comment that "any scale" means something like "laptop size to PB datasets processed on thousands of cores"? Makes this sentence long, but I feel it would be useful. We go into this later, though.

jbednar replied:
Your call on that one -- if you think later is too late and skimming readers won't see it, we can put that info right here too.

martindurant replied:
"(laptop to supercomputer/datacentre)" ?

martindurant commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -38,13 +38,13 @@ \subsubsection{Domain independent}
 
 \subsubsection{Efficient}
-Because Python itself is a relatively slow, interpreted language, native Python code is often suitable only for relatively small problems, limiting the audience for any tool built purely in Python with Python dependencies. To make efficient use of the computational resources available, SOSA libraries either wrap fast code written in compiled languages like FORTRAN and C, or they use the Numba Pyt ...
+Because Python itself is a relatively slow, interpreted language, native Python code is often suitable only for relatively small problems, limiting the audience for any tool built purely in Python with Python dependencies. To make efficient use of the computational resources available, SOSA libraries either wrap fast code written in compiled languages like Fortran and C, or they use a Python comp ...
```
It's only Fortran if referring to >=f90 :)

jbednar replied:
I actually looked it up. :-) 
<img width="618" alt="image" src="https://github.com/scipy-conference/scipy_proceedings/assets/1695496/16ce6568-ff99-41d6-ad3b-4b0debc32625">

martindurant commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -121,7 +122,7 @@ \subsubsection{Domain-specific}
 \subsubsection{Visualization}
 
-Whether or not any domain-specific processing has been done, the raw and processed data then need to be visualized so that humans can understand and validate them. Most Python visualization tools (see pyviz.org for a complete list) have limitations that prevent them from being suitable for the SOSA stack. Many of them are limited to relatively small amounts of data, lack support for the various d ...
+Whether or not any domain-specific processing has been done, the raw and processed data then need to be visualized so that humans can understand and validate them. Most Python visualization tools (see \texttt{pyviz.org} for a complete list) have limitations that prevent them from being suitable for the SOSA stack. Many of them are limited to relatively small amounts of data, lack support for the  ...
```
You mean https://pyviz.org/tools.html ?

martindurant commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -131,13 +132,13 @@ \subsubsection{User Interface (UI)}
 
-A final category of tools underlying all the rest shown in figure \ref{sosa} centers on how these libraries are packaged and put together into Python environments for solving any particular problem. Many different options are available for Python package and environment management, but SOSA tools are typically used with the conda package and environment manager, because it tracks binary dependenc ...
-https://www.anaconda.com/blog/8-levels-of-reproducibility for details).
+A final category of tools underlying all the rest shown in figure \ref{sosa} centers on how these libraries are packaged and put together into Python environments for solving any particular problem. Many different options are available for Python package and environment management, but SOSA tools are typically used with the conda package and environment manager, because it tracks binary dependenc ...
```
citation here is for conda-project ? If not, a direct link would be nice.

martindurant commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -147,23 +148,25 @@ \subsection{History and background}
 \subsection{Alternatives to SOSA}
 
-The libraries in SOSA are of course not the only alternatives available in Python; each library individually has alternatives that have their own strengths and advantages. However, the alternative libraries have not (yet?) been integrated with SOSA tools, making it much more difficult to apply them to a SOSA-based project. For instance, Ray is an alternative approach to distributed computation th ...
+The libraries in SOSA are of course not the only alternatives available in Python; each library individually has alternatives that have their own strengths and advantages. However, the alternative libraries have not (yet?) been integrated with SOSA tools, making it much more difficult to apply them to a SOSA-based project. For instance, Ray \cite{moritz:arxiv18} is an alternative approach to dist ...
```
If you want to choose a different compute engine here, use Spark rather than Ray, because you can run dask workloads on ray, apparently quite well.

martindurant replied:
Correction: I see you poke spark specifically in the next paragraph. Maybe MPI or something else here?

jbednar replied:
Hmm. Does running a Dask workload on Ray give any advantages?  Normally I try not to call out alternatives unless they meet a need not already covered. And I think the point still stands that if you pick up Ray and work with it, you won't be able to use hvPlot to view your distributed data structure.

martindurant replied:
> Does running a Dask workload on Ray give any advantages?

According to the authors of dask-on-ray, yes - but I don't know how well this holds up.

> you won't be able to use hvPlot to view your distributed data structure.

If think you totally would, it amounts to setting the default dask scheduler, and any dask workload will be farmed to it.

martindurant commented on papers/james_bednar/james_bednar.tex:
```diff
@@ -147,23 +148,25 @@ \subsection{History and background}
+This paper is the first to describe SOSA as an entity or collection of tools. The specific tools we selected are those that in the opinion of the authors are well integrated with the other tools and provide a ``mix and match'' approach to putting together libraries to implement a particular analysis workflow. There is not currently any entity besides these authors and this paper that defines what ...
 
-Other relevant areas that need work include the development of domain-general tools for capturing metadata, conventions, consistent units, etc. per domain (like the CF conventions for climate science). Furthermore, improved tools for collaboration are needed, such as implementing parts of the JupyterHub roadmap (jupyterhub.readthedocs.io/en/stable/contributing/roadmap.html). Lastly, better docume ...
+Other relevant areas that need work include the development of domain-general tools for capturing metadata, conventions, consistent units, etc. per domain (like the CF conventions for climate science). Furthermore, improved tools for collaboration are needed, such as implementing parts of the JupyterHub roadmap (\texttt{jupyterhub.readthedocs.io/en/stable/ contributing/roadmap.html}). Lastly, bet ...
```
Extra space appeared in the jhub link

jbednar replied:
That space gives a line break with no hyphen; the URL is too long to fit on one line. This is a case where putting it in the bibliography would make sense for formatting, it just seems like such a strange thing to be referencing as anything other than a hyperlink.

deniederhut reviewed (approved):
> now including the word "battle". :-)

Haha thanks. This looks good to me

deniederhut commented:
@scoobies mark ready

jbednar commented:
The day after SciPy ended, the correct name of SOSA came to me: Pandata. It's obvious -- this project is explicitly the Pangeo.io stack, but generalized to cover any field of science, not specifically geoscience, leaving only the domain-independent portions (which is most of Pangeo!).  I had rejected "Panscience" as being inaccurate (since it's not only science, but engineering, finance, etc.), and had rejected "Panalysis" as being embarrassing, but since data is what unites all the fields where this stack applies... it's Pandata! Just as it does in "Pangeo", here the "pan" represents "all sizes of data", because it's a scalable stack. I've updated the github repo to https://github.com/pandata-org/pandata , and will need to update the paper now.

jbednar committed:
Updated name to Pandata

jbednar committed:
Updated name to Pandata

jbednar commented:
Ok, the paper has been updated to say Pandata throughout.

jbednar committed:
Cleanup pass

jbednar commented:
Should be ready for publishing in the proceedings now!

deniederhut reviewed (approved):
Thanks @LMedinaU and @tkoyama010 for helping us get this paper ready for publication! And thanks @jbednar for joining as SciPy this year! And also all the other things!

![giphy-1253698798](https://github.com/scipy-conference/scipy_proceedings/assets/6868396/5cd22af5-762b-4a3f-998d-577816ac2b29)

tkoyama010 commented:
![NASA LGTM](https://media.giphy.com/media/MUKtQ1aRXwkO8ntpfg/giphy.gif)
